tools/codechecker_report_hash/codechecker_report_hash/hash.py

```diff
@@ -0,0 +1,332 @@
+# -------------------------------------------------------------------------
+#                     The CodeChecker Infrastructure
+#   This file is distributed under the University of Illinois Open Source
+#   License. See LICENSE.TXT for details.
+# -------------------------------------------------------------------------
+""" CodeChecker hash generation algorithms. """
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+
+import hashlib
+import io
+import logging
+import os
+import plistlib
+import sys
+import traceback
+
+LOG = logging.getLogger('codechecker_report_hash')
+
+handler = logging.StreamHandler()
+formatter = logging.Formatter('[%(levelname)s] - %(message)s')
+handler.setFormatter(formatter)
+
+LOG.setLevel(logging.INFO)
+LOG.addHandler(handler)
+
+
+class HashType(object):
+    """ Report hash types. """
+    CONTEXT_SENSITIVE = 1
+    PATH_SENSITIVE = 2
+
+
+def __get_line(file_name, line_no, errors='ignore'):
+    """ Return the given line from the file.
+
+    If line_no is larger than the number of lines in the file then empty
+    string returns. If the file can't be opened for read, the function also
+    returns empty string.
+
+    Try to encode every file as utf-8 to read the line content do not depend
+    on the platform settings. By default locale.getpreferredencoding() is used
+    which depends on the platform.
+
+    Changing the encoding error handling can influence the hash content!
+    """
+    try:
+        with io.open(file_name, mode='r',
+                     encoding='utf-8',
+                     errors=errors) as source_file:
+            for line in source_file:
+                line_no -= 1
+                if line_no == 0:
+                    return line
+            return u''
+    except IOError:
+        LOG.error("Failed to open file %s", file_name)
+        return u''
+
+
+def __str_to_hash(string_to_hash, errors='ignore'):
+    """ Encodes the given string and generates a hash from it. """
+    string_hash = string_to_hash.encode(encoding="utf-8", errors=errors)
+    return hashlib.md5(string_hash).hexdigest()
+
+
+def remove_whitespace(line_content, old_col):
```

Comment:
I thinking to make this module usage a bit more strict and make every function which should not be used externally as a private function in the module.

Only a few functions should be available for the module user.
Maybe one function to generate a report hash from a diagnostic section:
`get_report_hash(diag or diag path, file_path/source file, hash_type)`
would call the these two module internal functions based on the value of the hash_type: `get_report_path_hash(diag['path'], file_path)` and `get_report_hash_path_sensitive(diag, source_file)` 

Another function to replace the hash in a plist file:
`use_report_hash(plist_file, hash_type=HashType.CONTEXT_SENSITIVE) or replace_report_hash(...)`

Every other function is module internal.

What do you think @csordasmarton ?